Event-Driven Gamepad Input API (#1000)
MicrosoftEdge/MSEdgeExplainers: GamepadButtonAndAxisEvents/explainer.md

Author: Open question: 
Should we consider adding rawgamepadinputchanged.getCoalescedEvents() when { coalesce = false } ? 
This would be similar in semantics and pattern to rawpointerupdate which is part of Pointer Events, see:
- https://developer.mozilla.org/en-US/docs/Web/API/Element/pointerrawupdate_event#example
- https://developer.mozilla.org/en-US/docs/Web/API/Element/pointerrawupdate_event#example

> # Gamepad Button and Axis Events

Reviewer: Please also update the root readme.md in MSEdgeExplainers to include this explainer in the active section.

https://github.com/MicrosoftEdge/MSEdgeExplainers?tab=readme-ov-file#active-explainers-

Author: Done, thanks

- This document status: **Active**
- Expected venue:**[W3C Web Applications Working Group](https://www.w3.org/groups/wg/webapps/)**
- Current version: **This document**

## Introduction

The current Gamepad API relies on continuous polling to detect input changes, which can lead to input latency and increased CPU usage. This proposal introduces a more efficient approach that combines event-driven input handling for button presses and releases with frame-based state consolidation, where a single event encapsulates all gamepad state changes that occur within an input frame.

- Event-Driven Input Handling: Immediate events (buttondown, buttonup) fire when discrete inputs like button presses or releases occur.

- Frame-Based Consolidation: gamepadchange event encapsulates all the gamepad state changes that occurred within an input frame, preventing event spam.

This proposal builds upon earlier work by Chromium developers while refining the hybrid model for improved performance.

## User-Facing Problem

> The Gamepad API does not natively support event-driven input handling, and applications must rely only on continuous polling for updates. This polling-based approach introduces input latency because scripts cannot sync their polling with the arrival of new input, leading to a delay between input detection and handling. If an application polls at a regular interval, the average added latency is half the polling interval. For example, polling at 60Hz results in a latency of about 8.33 milliseconds. While decreasing the polling interval can reduce this latency, it comes at the cost of increased CPU usage.

Author: The example is stated in the original proposal. 
https://github.com/w3ctag/design-reviews/issues/662
https://github.com/w3c/gamepad/pull/15

For the example: https://www.rtings.com/monitor/tests/motion/refresh-rate

### Goals

Reduce input latency by moving away from constant polling and introducing event-driven input handling. 

### Non-goals

The existing polling mechanism will not be deprecated at all. We are just proposing an alternative way of handling input events and applications are free to select whichever they prefer.

> ## User research

Reviewer: I think we can cut the user research section.  Let's focus on our proposed implementation instead of Firefox's old prototype.

> Research and experimental implementations in Firefox suggest that event-driven input handling can significantly improve responsiveness and efficiency. Based on this research, the following challenges with prior implementations have been identified and addressed:

Author: No I don't have access to Firefox's implementation, I'm inferring this from Google's proposal.

> Research and experimental implementations in Firefox suggest that event-driven input handling can significantly improve responsiveness and efficiency. Based on this research, the following challenges with prior implementations have been identified and addressed:

Reviewer 2: I think it is fine. Just asking because other people might ask this as well. Firefox is open-source btw: https://github.com/mozilla/gecko-dev.

- Unintuitive event targets: Previous event systems fired on the global window object, which was counterintuitive. The windows object is not directly tied to a specific gamepad instance and since multiple gamepads can be connected, handleing events at the gamepad level will be more logical. Most input APIs like KeyboardEvent and MouseEvent dispatch events on the relevant input device not the global window. Gamepad events on window break this consistency. With the new proposal, events will be triggered directly on the Gamepad object that generated the event, providing a clearer and more developer-friendly experience.

- Event spam from per-frame axis updates:When a separate event for each changed axis on each frame of input is fired, if events are received too rapidly then they will queue, introducing input latency. 

> ## Proposed Approach

Reviewer: We need to add sample code that shows usage of the new APIs.  The current examples do not show how web developers would use these APIs.

For example:

https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/IndexedDbGetAllEntries/explainer.md#key-scenarios

See also:

Writing Effective Explainers: https://w3ctag.org/explainers/, which says "Practical Use Cases: Show example pieces of code that use the proposed solution to solve the user-facing problems."

To address the challenges of handling gamepad input efficiently while ensuring real-time responsiveness, an approach is proposed to event-driven input handling along with frame-based state consolidation.This proposal would add four new events that fire on the Gamepad object:

1) buttonup: These events fire only when a button is pressed (pressed attribute changes to false).

2) buttondown: These events fire only when a button is released (pressed attribute changes to true).

If multiple input frames have been received since the last time the event listener was invoked then the event listener is invoked only once with the most recent data received from the device. The getIntermediateEvents method returns the list of event objects representing the intermediate events that were not dispatched. An application only needs to subscribe to the types of events it is interested in handling.

```js
// Example Event Code

// When a button is pressed.
buttondown {
  "type": "buttondown",
  "gamepadIndex": 0,
  "buttonIndex": 2,
  "buttonSnapshot": { "pressed": true, "touched": true, "value": 1 },
  "gamepadTimestamp": 1234.567
}

// When a button is released.

> buttonup {

Reviewer 2: Up for discussion, but I think that `buttonchanged` could be better. With discrete buttons - only 2 states (pressed and released) - it is simple to implement. However, with analog buttons, how do you determine when to fire `buttondown` and `buttonup`? For example:
1. my left trigger (LT) is in the resting position, so it should report `0.0`. 
2. I press it halfway and it should report `0.5`. A buttondown event is fired.
3. I depress it so it now reports `0.25`. 
 
What should we do? The trigger is still pressed, but I released a little. Should we fire `buttondown`? Should we fire a `buttonup` with the update value? If we decide to fire a `buttondown` event, we need to track internally the previous value to calculate the negative delta, which adds complexity.

Author: Button up and button down events are best suited for digital buttons, where the state is either pressed or released. For analog buttons, a buttonchange event can be used to trigger whenever the value changes. This means that if the value increases from 0.25 to 0.5, an event should fire, and if it decreases again, another event should fire to reflect the change.

Reviewer 2: But there is no `buttonchange` event in your explainer. I am suggesting that it may be better to merge `buttonup` and `buttondown` into a new type of event called `buttonchanged`.

Reviewer: +1.  See Matt's comment here:

https://docs.google.com/document/d/1rnQ1gU0iwPXbO7OvKS6KO9gyfpSdSQvKhK9_OkzUuKE/edit?pli=1&disco=AAAAPRco2hU

"After discussion on the TAG review I'm leaning towards removing the buttondown/buttonup events. In the current proposal, gamepadchange would fire whenever buttondown or buttonup would fire, and the state contained by gamepadchange is a superset of the button events, so there's little benefit in keeping them."

Also, read through this discussion here:

https://github.com/w3ctag/design-reviews/issues/662

I think we want to focus on the rawgamepadchange event only to minimize latency for our partners.

Author: Button up and button down events are best suited for digital buttons, where the state is either pressed or released. For analog buttons, a buttonchange event can be used to trigger whenever the value changes. This means that if the value increases from 0.25 to 0.5, an event should fire, and if it decreases again, another event should fire to reflect the change.

  "gamepadTimestamp": 1234.567
}

```
```
// Proposed IDL for GamepadButtonEvent

interface GamepadButtonEvent : Event {
    readonly attribute long gamepadIndex;
    readonly attribute long buttonIndex;
    readonly attribute GamepadButton buttonSnapshot;
    readonly attribute DOMHighResTimeStamp gamepadTimestamp;

    sequence<GamepadButtonEvent> getIntermediateEvents();
};

```
3) gamepadchange event: Will also fire on the Gamepad object, and will include the snapshot of all the changes (buttons and axes) that happened within that input frame. The axesChange and buttonsChanged arrays contain indices of the axes and buttons for which the value attribute changed in the input frame, and the buttonsPressed and buttonsReleased arrays contain indices of buttons for which the pressed attribute changed in the input frame.

The getCoalescedEvents() method is used to return a sequence of events that have been coalesced (combined) together. 

How it woriks: 
To avoid firing too many events in quick succession for performance issues, the browser may choose to delay firing the gamepadchange event. When this happens, the browser adds the event to an internal queue.

> Before firing a buttondown or buttonup event (indicating a button has been pressed or released) or before running animation callbacks (e.g., requestAnimationFrame), the event queue is flushed. This means that all the events that have been delayed will be combined into one single event, representing the union of all changes up to that point.

Reviewer 2: What should show up in `gamepadchange` if I press and release the same button during an input frame?

Author: in that case the final state of the button is captured. if button 0 is pressed and released and pressed again and not released for example, then: 
buttonsChanged: [0],   
buttonsPressed: [0],
buttonsReleased: []

  buttonsChanged: [0],
  buttonsPressed: [0],
  buttonsReleased: [],
}
```
```
// Web IDL for the gamepadchange and rawgamepadchange event.

interface GamepadChangeEvent : Event {
  const DOMString type = "gamepadchange";
  readonly attribute Gamepad gamepadSnapshot;
  readonly attribute FrozenArray<long> axesChanged;
  readonly attribute FrozenArray<long> buttonsChanged;
  readonly attribute FrozenArray<long> buttonsPressed;
  readonly attribute FrozenArray<long> buttonsReleased;

  sequence<GamepadChangeEvent> getCoalescedEvents();
};

interface RawGamepadChangeEvent : Event {
  const DOMString type = "rawgamepadchange";
  readonly attribute Gamepad gamepadSnapshot;
  readonly attribute FrozenArray<long> axesChanged;
  readonly attribute FrozenArray<long> buttonsChanged;
  readonly attribute FrozenArray<long> buttonsPressed;
  readonly attribute FrozenArray<long> buttonsReleased;
};

```

### Dependencies on non-stable features

None identified at this stage. The proposal builds upon the existing Gamepad API.

## Alternatives considered

### Alternative 1: axischange and buttonchange events

axischange event: that fires when a member of the Gamepad's axes array changes.

> buttonchange event: that fires when a GamepadButton's value attribute changes.

Reviewer 2: Isn't `buttondown` and `buttonup` basically `buttonchange`? Maybe we should just keep `axischange` here.

Author: From my understanding of the "original proposal," the buttonchange event was intended for analog buttons, where the value attribute could change incrementally by decimal values. However, both buttonchange and axischange events were generating too many events, leading to event spam. To address this, intermediate events that didn’t need immediate dispatch were captured in a list using getIntermediateEvents. The gamepadchange event was then proposed as a solution to manage this excess event generation. Meanwhile, in the buttondown and buttonup events, only the pressed attribute changes, with the value toggling strictly between 0 and 1.

- Event Overload: A gamepad with multiple axes and analog buttons may trigger many events within a single input frame, especially if several axis or button values are modified at the same time.
- Sequential Input Processing: Applications may prefer to process related inputs simultaneously rather than one after the other. For example, both the X and Y axes of a thumbstick are usually handled together.

## Accessibility, Privacy, and Security Considerations
To prevent abuse and fingerprinting:
User Gesture Required:  Gamepad events won’t start firing until a user interacts with the gamepad (e.g., pressing a button). [Gamepad user gesture](https://www.w3.org/TR/gamepad/#dfn-gamepad-user-gesture)

Limit Persistent Tracking (fingerprinting): gamepadchange event will not expose device-specific identifiers. By default, no gamepad state is exposed to the tab even when gamepads are connected. [Fingerprinting in Web](https://www.w3.org/TR/fingerprinting-guidance/)

## Stakeholder Feedback / Opposition
Firefox: [#554 Gamepad API button and axis event](https://github.com/mozilla/standards-positions/issues/554)

> WebKit JS: Positive

Reviewer 2: What is "WebKit JS"? For these positions, basically what I have observed is that Firefox=Mozilla and Safari=WebKit.

Safari: No Signal

Web Developers: Positive

## References & acknowledgements

Thanks to the following contributors and prior work that influenced this proposal:

Firefox’s experimental implementation

Chromium Prior discussions on improving gamepad input handling.

Many thanks for valuable feedback and advice from:
- Gabriel Brito
- Matt Reynolds
- Steve Becker

Thanks to the following proposals, projects, libraries, frameworks, and languages
for their work on similar problems that influenced this proposal.

> - [Original Proposal](https://docs.google.com/document/d/1rnQ1gU0iwPXbO7OvKS6KO9gyfpSdSQvKhK9_OkzUuKE/edit?pli=1&tab=t.0)

Reviewer: nit: Please link this where you discuss Chrome's prior discussion above.

# Gamepad Button and Axis Events

## Authors:

- Sneha Agarwal (https://github.com/snehagarwal_microsoft)
- Steve Becker (https://github.com/SteveBeckerMSFT)
- Gabriel Brito (https://github.com/gabrielsanbrito)

## Participate
- [Issue tracker]
- [Gamepad API input events #662](https://github.com/w3ctag/design-reviews/issues/662)

## Status of this Document

This document is a starting point for engaging the community and standards bodies in developing collaborative solutions fit for standardization. As the solutions to the problems described in this document progress along the standards-track, we will retain this document as an archive and use this section to keep the community up-to-date with the most current standards venue and content location of future work and discussions.

- This document status: **Active**
- Expected venue:**[W3C Web Applications Working Group](https://www.w3.org/groups/wg/webapps/)**
- Current version: **This document**

## Introduction

> The current Gamepad API relies on continuous polling to detect input changes, which can lead to input latency and increased CPU usage. This proposal suggests using a single event API that is fired as soon as an input change (button and axis) is detected. If multiple changes happen in a rapid succession, each change triggers a separate event instead of merging these changes in to a single event.

Author: That's correct. In this version there is no coalescing.

This proposal builds upon earlier work by Chromium developers for improved performance, [Original Proposal](https://docs.google.com/document/d/1rnQ1gU0iwPXbO7OvKS6KO9gyfpSdSQvKhK9_OkzUuKE/edit?pli=1&tab=t.0)

## User-Facing Problem

> The Gamepad API lacks event-driven input handling, forcing applications to rely on continuous polling to query for changes in gamepad input. The continuous polling introduces input latency, as scripts cannot synchronize with new input fast enough. If the script increases the polling frequency to mitigate input latency, it increases CPU usage, impacting efficiency and battery life of the device.

Reviewer 2: I also think that we should refrain from mentioning these "hardware performance" points. Also the one in the paragraph below where we claim that it will improve battery usage.

> The Gamepad API lacks event-driven input handling, forcing applications to rely on continuous polling to query for changes in gamepad input. The continuous polling introduces input latency, as scripts cannot synchronize with new input fast enough. If the script increases the polling frequency to mitigate input latency, it increases CPU usage, impacting efficiency and battery life of the device.

Author: @gabrielsanbrito I couldn’t find that specific claim. I think what you are referring to is cause and effect. Since battery usage is a known concern from our partners, it made sense to include that data. But I get your point. I have updated this section to focus on input latency. Thanks

> The Gamepad API lacks event-driven input handling, forcing applications to rely on continuous polling to query for changes in gamepad input. The continuous polling introduces input latency, as scripts cannot synchronize with new input fast enough. If the script increases the polling frequency to mitigate input latency, it increases CPU usage, impacting efficiency and battery life of the device.

Author: @SteveBeckerMSFT When you posted this comment, I was initially under the impression that an event would be fired for every individual movement as it occurred. But that’s not actually the case.

The current understanding is that each event corresponds to an input frame, which is derived from a HID report is essentially a full snapshot of the gamepad’s state at a given point in time. Rather than emitting separate events for each button or axis change, the browser would dispatch a single event per input frame, containing all the relevant input data. This approach should help avoid an excessive number of events being generated.

### Goals

Reduce input latency by moving away from constant polling and introducing event-driven input handling. 

### Non-goals

The existing polling mechanism will not be deprecated. We are just proposing an alternative way of handling input events and applications are free to select whichever they prefer.

## Proposed Approach
To address the challenges of input latency, this proposal introduces a new event-driven mechanism: The `rawgamepadinputchange` event. This event fires directly on the [Gamepad](https://w3c.github.io/gamepad/#dom-gamepad) object and delivers real-time updates for each input change, eliminating the need for high-frequency polling. The `rawgamepadinputchange` event includes detailed information about the state of the gamepad at the moment of change:

- `axesChanged` and `buttonsChanged`: Arrays of indices indicating which axes or button values changed since the last event.

- `buttonsPressed` and `buttonsReleased`: Indices of buttons whose pressed state transitioned (from pressed to released or vice versa).

- `gamepadSnapshot`: A complete snapshot of the gamepad's current state, including all axes, buttons, ID, index, and timestamp.

> A new `rawgamepadinputchange` event is dispatched for every individual input state change, without delay or coalescing, enabling latency-sensitive applications—such as rhythm games, cloud gaming, or real-time multiplayer scenarios—to respond immediately and accurately to input.

Author: As discussed offline we decided to layer that in later. We can proceed with the rawgamepadinputchange event without the coalescing option to begin with.

## Alternatives considered
gamepadchange event: Similar to `rawgamepadinputchange` event but instead the getCoalescedEvents() method is used to return a sequence of events that have been coalesced (combined) together.

How it works: To avoid firing too many events in quick succession for performance issues, the browser may choose to delay firing the gamepadchange event. When this happens, the browser adds the event to an internal queue.

Before running animation callbacks (e.g., requestAnimationFrame), the event queue is flushed. This means that all the events that have been delayed will be combined into one single event, representing the union of all changes up to that point.

The final, combined gamepadchange event represents the combined state of the gamepad from all the events that were delayed and coalesced. When this event is dispatched, it contains all the changes that occurred during the delayed period.

## Accessibility, Privacy, and Security Considerations
To prevent abuse and fingerprinting, a ["gamepad user gesture"](https://www.w3.org/TR/gamepad/#dfn-gamepad-user-gesture) will be required before `RawGamepadInputChange` events start firing (e.g., pressing a button).

> Limit Persistent Tracking (fingerprinting): `rawgamepadinputchange` event will not expose device-specific identifiers. By default, no gamepad state is exposed to the tab even when gamepads are connected. [Fingerprinting in Web](https://www.w3.org/TR/fingerprinting-guidance/)

Reviewer: nit: can we simplify "Limit Persistent Tracking" by stating that the event will not expose any new state that is not already exposed by polling?

## Stakeholder Feedback / Opposition

> Firefox: [#554 Gamepad API button and axis event](https://github.com/mozilla/standards-positions/issues/554)

Reviewer: Is FireFox considered "No signal" since there is no response to the position linked above.

- Expected venue:**[W3C Web Applications Working Group](https://www.w3.org/groups/wg/webapps/)**
- Current version: **This document**

## Introduction

The current Gamepad API relies on continuous polling to detect input changes, which can lead to input latency and increased CPU usage. This proposal suggests using a single event API that is fired as soon as an input change (button and axis) is detected. If multiple changes happen in a rapid succession, each change triggers a separate event instead of merging these changes in to a single event.

This proposal builds upon earlier work by Chromium developers for improved performance, [Original Proposal](https://docs.google.com/document/d/1rnQ1gU0iwPXbO7OvKS6KO9gyfpSdSQvKhK9_OkzUuKE/edit?pli=1&tab=t.0)

## User-Facing Problem

The Gamepad API lacks event-driven input handling, forcing applications to rely on continuous polling to query for changes in gamepad input. The continuous polling introduces input latency, as scripts cannot synchronize with new input fast enough. If the script reduces the polling interval to mitigate input latency, it increases CPU usage, impacting efficiency and battery life of the device.

> This issue is particularly problematic for Xbox cloud gaming (xCloud), which streams Xbox games to a browser and depends on real-time gamepad input. To minimize latency, they currently poll every 4ms, but this high-frequency polling increases CPU usage and battery drain, especially on laptops and mobile devices. Additionally, since the Gamepad API is only supported on the main UI thread, it causes thread contention, particularly on low-end devices. In some cases, these constraints force them to reduce polling frequency, increasing input latency as a trade-off.

Author: Sure, thanks, done.

A more efficient solution would be an event-driven Gamepad API, similar to mouse and keyboard events, enabling real-time responsiveness without the overhead of constant polling. 

### Goals

Reduce input latency by moving away from constant polling and introducing event-driven input handling. 

### Non-goals

The existing polling mechanism will not be deprecated. We are just proposing an alternative way of handling input events and applications are free to select whichever they prefer.

## Proposed Approach
To address the challenges of input latency an approach is proposed to event-driven input handling. This proposal would add one new event that fires on the Gamepad object:

> rawgamepadinputchange event fires on the Gamepad object and delivers a snapshot of all gamepad input changes (axes and buttons) that occurred since the last event.

Author: Done, thanks.

```js
// Example rawgamepadinputchange Event:
rawgamepadinputchange {
  type: "rawgamepadchange",
  gamepadSnapshot: Gamepad {
    id: "Xbox Wireless Controller (STANDARD GAMEPAD Vendor: 045e Product: 02fd)",
    index: 0,
    connected: true,
    mapping: "standard",
    buttons: [
      // index 0 - button A
      { pressed: true, touched: false, value: 1.0 },
      // index 1 - button B
      { pressed: false, touched: true, value: 0.0 },
      // index 2 - analog button
      { pressed: false, touched: false, value: 0.5 },
      ...
    ],
    // [left stick X, left stick Y, right stick X, right stick Y]
    axes: [0.25, -0.5, 0.0, 0.0],
    timestamp: 9123456.789
  },
  // Left stick X and Y moved since last event.
  axesChanged: [0, 1],
  // button index 0 was pressed, button index 2 value changed
  buttonsValueChanged: [0, 2],
  // button index 0 pressed
  buttonsPressed: [0],
  // button index 1 was touched
  buttonsTouched: [1],
  // High-res timestamp when change was detected
  gamepadTimestamp: 9123456.789 
}
```
## Proposed IDL
```

> [Exposed=Window, SecureContext]

Reviewer 2: The Gamepad interface does not require a `SecureContext` anymore: https://www.w3.org/TR/gamepad/#gamepad-interface

Author: Thanks. removed

interface Gamepad {
  // New attributes
  attribute EventHandler onrawgamepadinputchange;

  // Existing attributes
  readonly attribute DOMString id;
  readonly attribute long index;
  readonly attribute boolean connected;
  readonly attribute DOMHighResTimeStamp timestamp;
  readonly attribute GamepadMappingType mapping;
  readonly attribute FrozenArray<double> axes;

>   readonly attribute FrozenArray<GamepadButton> buttons;

Author: Thanks, Done!

};
```
### RawGamepadInputChangeEvent interface IDL, used for rawgamepadinputchange.
```
[Exposed=Window, SecureContent]

> interface RawGamepadInputChangeEvent : Event {

Author: GamepadEvent is tailored for connection events, and this event has a different purpose and additional fields. I think we should continue to inherit from events to get all the standard behaviors (type, bubble, cancelable, etc).

> interface RawGamepadInputChangeEvent : Event {

Reviewer 2: I couldn't find anything in the `GamepadEvent` interface description that suggests that is tailored for connection events. `GamepadEvent` is just a generic event that provides a gamepad snapshot. Also, `GamepadEvent` is an `Event`, so I don't understand why would we lose the event behaviors you mentioned.

>   constructor(DOMString type, optional RawGamepadInputChangeEventInit eventInitDict = {});

Reviewer 2: I just realized that you didn't add a definition for `RawGamepadInputChangeEventInit`. We should add it too.

    for (let buttonIndex of changeEvent.buttonsValueChanged) {
      const buttonValue = changeEvent.gamepadSnapshot.buttons[buttonIndex].value;
      console.log('button ' + buttonIndex +
                  ' on gamepad ' + changeEvent.gamepadSnapshot.index +
                  ' changed to value ' + buttonValue);
    }

    // Binary buttons pressed
    for (let buttonIndex of changeEvent.buttonsPressed) {
      const buttonPressed = changeEvent.gamepadSnapshot.buttons[buttonIndex].pressed;
      console.log('button ' + buttonIndex +
                  ' on gamepad ' + changeEvent.gamepadSnapshot.index +
                  ' changed pressed to ' + buttonPressed);
    }

    // Buttons touched
    for (let buttonIndex of changeEvent.buttonsTouched) {
      const buttonTouched = changeEvent.gamepadSnapshot.buttons[buttonIndex].touched;
      console.log('button ' + buttonIndex +
                  ' on gamepad ' + changeEvent.gamepadSnapshot.index +
                  ' changed touched to ' + buttonTouched);
    }
  };
};

```

### Dependencies on non-stable features

None identified at this stage. The proposal builds upon the existing Gamepad API.

## Alternatives considered

> gamepadchange event: Similar to rawgamepadinputchange event but instead the getCoalescedEvents() method is used to return a sequence of events that have been coalesced (combined) together.

Reviewer 2: Maybe add more context about it? Like the proposed IDL and a link to where this was proposed and discussed?

How it works: To avoid firing too many events in quick succession for performance issues, the browser may choose to delay firing the gamepadchange event. When this happens, the browser adds the event to an internal queue.

Before running animation callbacks (e.g., requestAnimationFrame), the event queue is flushed. This means that all the events that have been delayed will be combined into one single event, representing the union of all changes up to that point.

The final, combined gamepadchange event represents the combined state of the gamepad from all the events that were delayed and coalesced. When this event is dispatched, it contains all the changes that occurred during the delayed period.

## Accessibility, Privacy, and Security Considerations

> To prevent abuse and fingerprinting:
> User Gesture Required:  Gamepad events won’t start firing until a user interacts with the gamepad (e.g., pressing a button). [Gamepad user gesture](https://www.w3.org/TR/gamepad/#dfn-gamepad-user-gesture)

Reviewer 2: ```suggestion
To prevent abuse and fingerprinting, a ["gamepad user gesture"](https://www.w3.org/TR/gamepad/#dfn-gamepad-user-gesture) will be required before `RawGamepadInputChange` events start firing (e.g., pressing a button).
```

> Limit Persistent Tracking (fingerprinting): gamepadchange event will not expose device-specific identifiers. By default, no gamepad state is exposed to the tab even when gamepads are connected. [Fingerprinting in Web](https://www.w3.org/TR/fingerprinting-guidance/)

Reviewer 2: Is it valid? We don't have `gamepadchange` events anymore. Also the `Gamepad::id` in the gamepad snapshot will contain device-specific identifiers - e.g. PID/VID pair.

Author: Device specific id's reveal the make and model of the controller but does not uniquely identify a user.

## Stakeholder Feedback / Opposition
Firefox: [#554 Gamepad API button and axis event](https://github.com/mozilla/standards-positions/issues/554)

Safari: No Signal

> xCloud: Positive

Author: Done

By integrating these techniques, this approach minimizes CPU overhead while ensuring low-latency input processing, providing developers with a more efficient and flexible way to handle gamepad interactions. This proposal builds upon earlier work by Chromium developers while refining the hybrid model for improved performance.

## User-Facing Problem

The Gamepad API does not natively support event-driven input handling, and applications must rely on continuous polling for updates, which:

- Introduces input latency, as polling intervals may not align with the precise timing of input changes.

- Increases CPU usage, particularly when polling at high frequencies.

- Lacks efficiency in handling gamepad input changes, as it does not account for meaningful input variations, leading to unnecessary events being triggered.

### Goals

1) **Enhance gamepad input handling** by transitioning from continuous polling to event-driven mechanisms, improving responsiveness.

> 2) **Optimize performance** by reducing input latency and minimizing CPU overhead through a hybrid approach of event handling and threshold-based polling.

Reviewer 2: > reducing input latency

I am failing to see how this proposal reduces input latency. In the backend we are still polling and in one part of the document we even proposed reducing the polling interval.

> 3) **Provide flexibility for developers** by allowing efficient management of gamepad input events, enabling granular control over when and how input changes are processed.

Reviewer 2: I am not sure that I understand what this sentence means. What "efficient management" and "granular control" the current API does not offer?

### Non-goals

> Deprecating the existing polling mechanism: The polling mechanism will not be fully deprecated, as it is still necessary for detecting small input changes that events alone may not address.

Reviewer 2: > The polling mechanism will not be fully deprecated

Make it clear that the intention of this proposal is not to the deprecate polling at all. We are just proposing an alternative way of handling input events and applications are free to select whichever they prefer.

Polling Intervals: Polling will occur at controlled intervals (e.g., every 50ms) instead of continuously on every frame. This reduces CPU overhead by avoiding constant polling when there is no meaningful change.

### Solving goal 3 with this approach.
Goal: Provide Flexibility for Developers to Efficiently Manage Gamepad Input Events.

Customizable Event Handling: Developers can set their own threshold values for button or axis changes, enabling different types of input handling (e.g., more sensitive input for precise games like racing sims, or less sensitivity for more casual games).
Developers can decide how to handle different gamepad states (pressed, released, touched) through event detail properties such as buttonSnapshot or axisSnapshot.

Developer-Defined Event Handling: By using events like gamepadchange, developers can register event listeners to define exactly how gamepad input should be processed, such as applying force feedback when a button is pressed or initiating an action when a specific button combination is detected.

> Access to Raw Data: The game developer can access both raw and threshold-processed data (e.g., axis values and button states) to design input controls that suit the needs of the game, such as adjusting for gamepad types, handling different input devices, or implementing specific interactions.

Reviewer 2: > Access to Raw Data: The game developer can access both raw and threshold-processed data (e.g., axis values and button states) to design input controls that suit the needs of the game, such as adjusting for gamepad types, handling different input devices, or implementing specific interactions.

Isn't this already provided?

### Alternative 3: getIntermediateEvents
Captures fine-grained input changes (such as axis movements or button states) that don't fit into the normal event cycle. Rather than having the browser fire events for every intermediate state, the game developer manually checks for significant changes and fires events when needed. The threshold mechanism replaces the need for getIntermediateEvents in this proposal.

## Accessibility, Privacy, and Security Considerations
To prevent abuse and fingerprinting:
User Gesture Required:  Gamepad events won’t start firing until a user interacts with the gamepad (e.g., pressing a button). [Gamepad user gesture](https://www.w3.org/TR/gamepad/#dfn-gamepad-user-gesture)

Limit Persistent Tracking (fingerprinting): gamepadchange event will not expose device-specific identifiers. By default, no gamepad state is exposed to the tab even when gamepads are connected. [Fingerprinting in Web](https://www.w3.org/TR/fingerprinting-guidance/)

## Stakeholder Feedback / Opposition
Firefox: Implemented experimental support for gamepadbuttondown and gamepadbuttonup, but identified issues with event spam and lost state.

> Chromium: Evaluating different approaches to event-driven gamepad input handling.[Original Proposal](https://docs.google.com/document/d/1rnQ1gU0iwPXbO7OvKS6KO9gyfpSdSQvKhK9_OkzUuKE/edit?pli=1&tab=t.0)

Reviewer 2: I don't think you need to add Chromium here (it's us and we are supportive). Instead you should add a line for WebKit (Safari)
